osd: Deny reservation if expected backfill size would put us over bac…

dzafman opened the pull request:
- [x] References tracker ticket
- [ ] ~~Updates documentation if necessary~~
- [x] Includes tests for new functionality or reproducer for bug

dzafman commented:
retest docs please

dzafman committed:
osd: Don't let values go negative, particularly num_bytes and num_objects

An ObjectStore corruption could cause an object to be larger causing its removal
to make num_bytes too small which could result in a negative value.

During scrub testing we deliberately corrupt objects, so we don't want
a very negative num_bytes that becomes a large positive value to cause
backfill too full, breaking the test.

Fixes: http://tracker.ceph.com/issues/24801

Signed-off-by: David Zafman <dzafman@redhat.com>

dzafman commented:
http://pulpito.ceph.com/dzafman-2018-08-31_11:29:01-rados-wip-zafman-testing-distro-basic-smithi/
2 failures - infrastructure
workloads/osd.yaml osd-backfill-space.sh FAILED (retested after correction)

http://pulpito.ceph.com/dzafman-2018-09-04_15:32:47-rados-wip-zafman-testing-distro-basic-smithi/
workloads/osd.yaml osd-backfill-space.sh PASSED

liewegas commented on src/osd/PG.cc:
```diff
@@ -6593,6 +6601,30 @@ void PG::_delete_some(ObjectStore::Transaction *t)
+    return std::max((int64_t)0, (bf_bytes >> 10) - (local_bytes >> 10));
+}
+
+int PG::pg_stat_adjust(osd_stat_t &new_stat)
```
can we pass a pointer instead of ref to communicate that the argument is modified?

dzafman commented:
@liewegas Fixed pg_stat_adjust to take pointer.

liewegas commented on src/osd/PG.cc:
```diff
@@ -6593,6 +6601,31 @@ void PG::_delete_some(ObjectStore::Transaction *t)
+    return std::max((int64_t)0, (bf_bytes >> 10) - (local_bytes >> 10));
+}
+
+int PG::pg_stat_adjust(osd_stat_t *ns)
```
This isn't protected by any lock. I think we should reuse pg_stats_publish_lock for this, and make sure on the PG side that we're updating primary_num_bytes under that lock?

dzafman replied:
The stat_lock is protecting the official osd_stat_t AND primary_num_bytes.

The only caller OSDService::compute_adjusted_ratio() uses the result to compute the adjusted ratio.  The osd_stat_t in compute_adjusted_ratio() is a local copy from its callers OSDService::_tentative_full() and OSD::heartbeat().  OSD::heartbeat() uses set_osd_stat() that locks with stat_lock before returning a copy of a modified osd_stat_t.  OSDService::_tentative_full() caller PG::RecoveryState::RepNotRecovering::react(const RequestBackfillPrio &evt) uses the stat_lock to not only retrieve the osd_stat_t, but also to set primary_num_bytes at the point were the decision to go forward with the backfill has been made.  Once primary_num_bytes is set we can always adjust the ratio based in progress backfill.

liewegas replied:
I'm thinking of the PG members primary_num_bytes and info members used below, which are updated under the PG lock.  I think we add the PG::pg_stats_publish_lock around those updates, and take that lock here for each PG we examine, it will be safe?

dzafman commented:
@liewegas I think I have a better way to handle the num_bytes and primary_num_bytes that are sometimes examined and modified without the pg lock.  Using atomic<> should be more efficient then a Mutex and modify a memory location.  Plus I wasn't thrilled using pg_stats_publish_lock already being used for something else.  There is an atomic PG::num_bytes that tracks (under pg lock) with info.stats.stats.sum.num_bytes.  It doesn't matter if a thread is modifying atomic num_bytes with pg lock with another thread retrieving it to do a backfill reservation without pg lock.

The state machine prevents contention between with setting primary_num_bytes > 0 and resetting to 0.  It only gets set to a value > 0 with stat_lock because it represents a reservation of bytes pending from backfill.  I'm thinking of renaming it to reserved_num_bytes or backfilling_num_bytes.

liewegas commented:
+1 on renaming to reserved_num_bytes

I'm worried about the num_bytes atomic because it can easily fall out of sync with the info.stats field. Can we instead make an accessor (get_backfilled_num_bytes()) that takes the appropriate locks (stats lock)?  Or is the problem that when the stats are updated on the replica the stats lock isn't taken (or "published")?

dzafman commented:
@liewegas This is the easiest version to maintain which just uses the PG lock very briefly to grab the PG::info.stats.stats.sum.num_bytes.  It is already updated in all cases with the PG lock.

I'll create an example branch which has to get the pg_stats_publish_lock when modifying the info.stats.stats.sum.  The difficulty is that pg_info_t is used in places other than the PG, but only in a PG context and only when modifying info.stats.stats.sum is the extra lock needed.

dzafman commented:
@liewegas I realize now that the PG is already locked in the state machine.  So the only locking issue left is in PG::pg_stat_adjust() which I've addressed.  I think that with the atomic<> for reserved_num_bytes and 1 case of grabbing the PG lock, that is rare, we don't need any other locking.

The other PG lock is grabbed in PG::get_stats_num_bytes() which is only used in testing through a config variable.

dzafman commented:
All failure unrelated to this change
http://qa-proxy.ceph.com/teuthology/dzafman-2018-09-12_18:57:34-rados-wip-zafman-testing-distro-basic-smith
	326 PASSED
	3012418 FAILED test_envlibrados_for_rocksdb.sh missing header causes test compilation failure
	3012505 FAILED LibRadosTwoPoolsPP.ManifestUnset (included chance_pgnum_shrink: 1)
    	3012467 DEAD Infrastructure

liewegas commented:
I see a deadlock with pg->lock vs osd->stat_lock:
```
pg is locked
PG::RecoveryState::RepNotRecovering::react(const RequestBackfillPrio &evt)
        lock pg->osd->stat_lock
        pg->osd->tentative_backfill_full
                _tentative_full
                        compute_adjusted_ratio
                                pg_stat_adjust
                                        pg lock *****
pg is locked
PG::RecoveryState::RepNotRecovering::react(const RequestBackfillPrio &evt)
        lock pg->osd->stat_lock
        pg->osd->tentative_backfill_full
                _tentative_full
                        compute_adjusted_ratio
                                pg_stat_adjust
                                        pg lock *****
```
I think the problematic area is probably where the pg code has to call out to an OSDService helper that then calls into every single PG.  either we need to use a nested lock (pg_stat_lock?) or avoid that O(n) check?

dzafman commented:
@liewegas We have 2 asserts that verify that the deadlock can't happen.  In PG::RecoveryState::RepNotRecovering::react(const RequestBackfillPrio &evt) we have #1: assert(pg->get_reserved_num_bytes() == 0); "   That is the first pg that is locked.  Later in pg_stat_adjust() we have #2: assert(!_lock.is_locked_by_me());" which must be true since we only lock a pg that reserved_num_bytes != 0.  pg_stat_adjust() only locks PGs that are already backfilling.

dzafman committed:
test: Add testing for backfill out of space detection

Signed-off-by: David Zafman <dzafman@redhat.com>

dzafman committed:
osd: Deny reservation if expected backfill size would put us over backfill_full_ratio (EC)

Erasure Coded Pools

Fixes: http://tracker.ceph.com/issues/19753

Signed-off-by: David Zafman <dzafman@redhat.com>

dzafman committed:
osd: Deny reservation if expected backfill size would put us over backfill_full_ratio

Replicated pools

Fixes: http://tracker.ceph.com/issues/19753

Signed-off-by: David Zafman <dzafman@redhat.com>

dzafman committed:
osd: Remove redundant backfill space check

Once we are reserving space we don't want a borderline case to be rejected
due to client activity taking us over the backfill full threshold once
we've committed to proceed.

Signed-off-by: David Zafman <dzafman@redhat.com>

dzafman committed:
test: Add singleton rados suite test for backfill full

This injects backfill full as opposed to lowering the backfill full ratio

Signed-off-by: David Zafman <dzafman@redhat.com>

dzafman committed:
test: Add testing for erasure code backfill out of space detection

Signed-off-by: David Zafman <dzafman@redhat.com>

dzafman commented:
TESTING PASSED
dzafman-2018-11-07_14:10:54-rados-wip-zafman-testing-22797-distro-basic-smithi
	3234519 valgrind memcheck error ceph-osd rocksdb?
	3234388 rebuild_mondb: Monitor daemon marked osd.0 down, but it is still running
	3234507 scrub_test: assert len(objs['inconsistents']) == 1
	3234615 osd-scrub-repair.sh TEST_corrupt_scrub_erasure_overwrites
	3234550 error: expected string-literal before ‘)’ token
	**3234564 osd-backfill-space.sh fixed and rerun below**
	3234472 mon_thrash with mon/test_mon_osdmap_prune.sh failure
	3234425 error: expected string-literal before ‘)’ token

A rerun with corrected test script only created 4 jobs and only 1 unrelated test failed
dzafman-2018-11-13_22:47:54-rados-wip-zafman-testing-22797-distro-basic-smithi
        3253770 /usr/include/rados/buffer.h:657:61: error: expected string-literal before ‘)’ token
standalone/osd was run and PASSED including osd-backfill-space.sh

dzafman commented:
New test rados/singleton/all/thrash-backfill-full.yaml did get run during the rados suite.

dzafman commented:
TESTING PASSED
http://pulpito.ceph.com/dzafman-2018-12-17_22:00:32-rados-wip-zafman-testing-distro-basic-smithi/
3373454 test_dump_pgstate_history: Can't find expected values in history object, failing
3373543 http://tracker.ceph.com/issues/37706
3373632 Infrastructure
3373649 FAIL: test_pool_list (tasks.mgr.dashboard.test_pool.PoolTest)
3373710 http://tracker.ceph.com/issues/37706

TESTING PASSED
run-standalone.sh run with fix for 37706

liuxiangsandstone commented on src/osd/OSD.cc:
```diff
@@ -684,20 +684,15 @@ float OSDService::get_failsafe_full_ratio()
-void OSDService::check_full_status(float ratio)
+OSDService::s_names OSDService::recalc_full_state(float ratio, float pratio, string &inject)
 {
-  std::lock_guard l(full_status_lock);
```
why？
